sqlite/ext/comdb2/systablepermissions.c

```diff
@@ -0,0 +1,291 @@
+    }
+    sqlite3_free(pCur->ppTables);
+    sqlite3_free(pCur);
+    return SQLITE_OK;
+}
+
+static int permissionsNext(sqlite3_vtab_cursor *cur)
+{
+    permissions_cursor *pCur = (permissions_cursor *)cur;
+    ++pCur->iUser;
+    if (pCur->iUser >= pCur->nUsers) {
+        pCur->iUser = 0;
+        ++pCur->iTable;
+    }
+    return SQLITE_OK;
+}
+
+static int permissionsColumn(sqlite3_vtab_cursor *cur, sqlite3_context *ctx,
+                             int i)
+{
+    permissions_cursor *pCur = (permissions_cursor *)cur;
+
+    tran_type *trans = curtran_gettran();
+    if (!trans) {
+        logmsg(LOGMSG_ERROR, "%s cannot create transaction object\n", __func__);
+        return 1;
+    }
+    char *tbl = pCur->ppTables[pCur->iTable].name;
+    int flags = pCur->ppTables[pCur->iTable].flags;
+    char *usr = pCur->ppUsers[pCur->iUser];
+    switch (i) {
+    case STTP_TABLE: {
+        sqlite3_result_text(ctx, tbl, -1, NULL);
+        break;
+    }
+    case STTP_USER: {
+        sqlite3_result_text(ctx, usr, -1, NULL);
+        break;
+    }
+    case STTP_READ:
+    case STTP_WRITE:
+    case STTP_DDL: {
+        int access, err;
+        if (i == STTP_READ) {
+            access = ACCESS_READ;
+        } else if (i == STTP_WRITE) {
+            access = ACCESS_WRITE;
+        } else {
+            access = ACCESS_DDL;
+        }
+        sqlite3_result_text(
+            ctx,
+            YESNO(((access == ACCESS_READ) && (flags & CDB2_ALLOW_ALL)) ||
+                  !bdb_check_user_tbl_access_tran(thedb->bdb_env, trans, usr,
+                                                  tbl, access, &err)),
+            -1, SQLITE_STATIC);
+        break;
+    }
+    }
+    curtran_puttran(trans);
+    return SQLITE_OK;
+}
+
+static int permissionsRowid(sqlite3_vtab_cursor *cur, sqlite_int64 *pRowid)
+{
+    permissions_cursor *pCur = (permissions_cursor *)cur;
+    *pRowid = pCur->iUser;
+    return SQLITE_OK;
+}
+
+static int permissionsEof(sqlite3_vtab_cursor *cur)
+{
+    permissions_cursor *pCur = (permissions_cursor *)cur;
+    if (pCur->nUsers == 0)
+        return 1;
+    return pCur->iTable >= pCur->nTables;
+}
+
+/*
+** This method is called to "rewind" the series_cursor object back
+** to the first row of output.  This method is always called at least
+** once prior to any call to xColumn() or xRowid() or xEof().
+*/
+static int permissionsFilter(sqlite3_vtab_cursor *pVtabCursor, int idxNum,
+                             const char *idxStr, int argc, sqlite3_value **argv)
+{
+    permissions_cursor *pCur = (permissions_cursor *)pVtabCursor;
+
+    pCur->iUser = 0;
+    pCur->iTable = 0;
+    return SQLITE_OK;
+}
+
+/*
+** There is no way to really take advantage of this at the moment.
+** The output of this table is a mostly unordered list of strings.
+*/
+static int permissionsBestIndex(sqlite3_vtab *tab, sqlite3_index_info *pIdxInfo)
+{
+    return SQLITE_OK;
+}
+
+const sqlite3_module systblSystablePermissionsModule = {
```

Comment:
The ezsystbl wrapper could save you a bit of coding.

Reply:
Indeed. The latest patch takes care of that.